jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostEstimator.scala

```diff
@@ -527,24 +530,54 @@ private[spark] trait XGBoostModel[M <: XGBoostModel[M]] extends Model[M] with ML
 
   /** Predict */
   private[spark] def predictInternal(booster: Booster, dm: DMatrix, pred: PredictedColumns,
-                                     batchRow: Iterator[Row]): Seq[Row] = {
-    var tmpOut = batchRow.toSeq.map(_.toSeq)
-    val zip = (left: Seq[Seq[_]], right: Array[Array[Float]]) => left.zip(right).map {
-      case (a, b) => a ++ Seq(b)
-    }
+                                     originalRowIter: Iterator[Row]): Iterator[Row] = {
+    val tmpIters: ArrayBuffer[Iterator[Row]] = ArrayBuffer.empty
     if (pred.predLeaf) {
-      tmpOut = zip(tmpOut, booster.predictLeaf(dm))
+      tmpIters += booster.predictLeaf(dm).map(Row(_)).iterator
     }
     if (pred.predContrib) {
-      tmpOut = zip(tmpOut, booster.predictContrib(dm))
+      tmpIters += booster.predictContrib(dm).map(Row(_)).iterator
     }
     if (pred.predRaw) {
-      tmpOut = zip(tmpOut, booster.predict(dm, outPutMargin = true))
+      tmpIters += booster.predict(dm, outPutMargin = true).map(Row(_)).iterator
     }
     if (pred.predTmp) {
-      tmpOut = zip(tmpOut, booster.predict(dm, outPutMargin = false))
+      tmpIters += booster.predict(dm, outPutMargin = false).map(Row(_)).iterator
+    }
+
+    // This is not so efficient considering that toSeq from first iterators will be called
+    // many times.
+    //    tmpIters.foldLeft(originalRowIter) { case (accIter, nextIter) =>
+    //      // Zip the accumulated iterator with the next iterator
+    //      accIter.zip(nextIter).map { case (a: Row, b: Row) =>
+    //        Row.fromSeq(a.toSeq ++ b.toSeq)
+    //      }
+    //    }
+
+    tmpIters.size match {
+      case 4 =>
+        originalRowIter.zip(tmpIters(0)).zip(tmpIters(1)).zip(tmpIters(2)).zip(tmpIters(3)).map {
```

Comment:
Could you please add some comments what you are doing here? This is not the easiest to read code with symbols `a,b,c,d,e` being concatenated.

Reply:
This piece of code is an optimization for 
``` scala
    // This is not so efficient considering that toSeq from first iterators will be called
    // many times.
    //    tmpIters.foldLeft(originalRowIter) { case (accIter, nextIter) =>
    //      // Zip the accumulated iterator with the next iterator
    //      accIter.zip(nextIter).map { case (a: Row, b: Row) =>
    //        Row.fromSeq(a.toSeq ++ b.toSeq)
    //      }
    //    }
```

to zip some iterators to a new iterator. For example

``` scala
iter1 = Seq("a", "b", c").iterator
iter2 = Seq("1", "2", "3").iterator
iter3 = Seq("xa", "xb", "xc").iterator
```

`iter1.zip(iter2).zip(iter3).map ...` could get a new iterator with content like 


``` scala
Seq(("a", "1", "xa"),   
        ("b", "2", "xb"),
        ("c", "3", "xc"))`
```

Reply:
That's unfortunate. Reminds me of lisp ` (car (cdr (car ...)))`. Are you certain that there's no way to make this more explicit (like using a helper function with a name).?

Reply:
No, we can't do that since we have no idea about the elements in the Array,  It could be originalRowIterator, rawPredictions, or originalRowIterator, leaf predictions.

Reply:
Would love to learn more about Scala one day, the macro implementation there looks interesting.